Add Unit Test for CompressOutbox to Tempfile Creation Refactoring (SOFTWARE-5540)

jeff-takaki opened the pull request:
@brianhlin I have been getting `AssertionError: False is not true` and I think it has to do with how I am using the mock patch. I tried a few different ways of incorporating the mock with not much success. It's possible I am provisioning the test environment incorrectly??

brianhlin committed:
Prep for sandbox refactoring with unit tests (SOFTWARE-5531)

brianhlin committed:
First attempt at cleaning up the tempfile creation (SOFTWARE-5540)

brianhlin committed:
First stab at updating temp tarball creation (SOFTWARE-5531)

jeff-takaki committed:
Add unit test for Compress Outbox function

brianhlin commented:
@jtakaki-matc I don't see anything obvious that would indicate why this failed so I would

1. Comment out the teardown function
2. Run this test locally and inspect the filesystem to see if the expected tarball gets created. This may be a good spot to add some changes to the actual `CompressOutbox` function to add debugging `print()` statements

jeff-takaki commented:
@brianhlin Test seems to be working, checked locally running from 2.x branch as well.

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,67 @@
+    @patch('gratia.common.config.ConfigProxy.get_GratiaExtension', create=True, return_value ='test-extension')
+    @patch('gratia.common.config.ConfigProxy.getFilenameFragment', create=True, return_value ='test-filename')
+    def test_compress_outbox(self, mock_gratia_ext, mock_file_frag):
+        """CompressOutbox compresses the files in the outbox directory
+            and stores the resulting tarball in probe_dir/staged.
+        """
```
```suggestion
        """CompressOutbox compresses the files in the outbox directory
        and stores the resulting tarball in probe_dir/staged.
        """
```

This multiline string is printed as is (try it in your python interpreter) so your extra 4 spaces will be included and look something like:

```
$ python
Python 3.11.4 (main, Jun  7 2023, 00:00:00) [GCC 12.3.1 20230508 (Red Hat 12.3.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print("""CompressOutbox compresses the files in the outbox directory
...     and stores the resulting tarball in probe_dir/staged.
... """)
CompressOutbox compresses the files in the outbox directory
    and stores the resulting tarball in probe_dir/staged.

```

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,67 @@
+
+        #Assert that CompressOutbox returns True
+        result = sandbox_mgmt.CompressOutbox(probe_dir, outbox, outfiles)
+        self.assertTrue(result)
```
A few weeks back we had a discussion about the different things that this function does: returns a boolean but also a tarball gets created in an expected location made up of the specified outfiles. It may be useful to test for things like:

1. The tarball is created in the expected location with expected pattern for the filename (similar to assertions in https://github.com/opensciencegrid/gratia-probe/pull/176/files#diff-a41929032101b1500faa4564484c19a2829fb12c86cd3b95182aa9a66e5071daR22-R29)
2. The unpacked tarball contents are the same as what's given in the outfiles

I'd probably organize these as separate tests and you could even refactor the call to the `CompressOutbox` function into the `setUp()` test!

brianhlin reviewed (changes requested):
This is looking great! One minor whitespace change and a request for additional tests.

EDIT: Could you also update the title of this PR? I don't think it summarizes ALL of the commits in the PR

jeff-takaki committed:
Remove excess whitespace comment

Co-authored-by: Brian Lin <brianhlin@gmail.com>

jeff-takaki committed:
Add Unit Tests for tarball creation and to verify tarball contents

jeff-takaki commented:
@brianhlin Added two more tests!

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+
+class CompressOutboxTests(unittest.TestCase):
+    def setUp(self):
+        #provision test environment
```
```suggestion
        # provision test environment
```

https://www.flake8rules.com/rules/E265.html. Make sure this gets applied to your other comments, too!

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertTrue(os.path.exists(path_to_tarball/{tarball}))
+
+    @patch('gratia.common.config.ConfigProxy.getFilenameFragment', create=True, return_value ='test-filename')
+    @patch('gratia.common.config.ConfigProxy.get_GratiaExtension', create=True, return_value ='test-extension')
```
Since you're applying these patches to every single test, you can refactor them out into `setUp()`. See https://github.com/opensciencegrid/gratia-probe/blob/2.x/test/test_condor.py#L126-L137 for an example

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        sandbox_mgmt.CompressOutbox(probe_dir, outbox, outfiles)
+
+        #Finds tarball that matches GenerateFilename() output
+        os.chdir(f'{probe_dir}/staged/store')
```
It's a little silly to define new variables for so few call sites. Just reference the `self.*` variables directly. That's what they're there for!

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        #Finds tarball that matches GenerateFilename function output
+        os.chdir(f'{probe_dir}/staged/store')
+        for tarball in glob.glob("tz.*.test-extension__*"): 
+            return(tarball)
```
Why return here? This will stop the test before your assertion below

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        for tarball in glob.glob("tz.*.test-extension__*"): 
+            return(tarball)
+        
+        self.assertTrue(os.path.exists(path_to_tarball/{tarball}))
```
This is the first reference to `path_to_tarball` that I can see so this will throw an exception. Also the `{}` formatting looks off. I'd suggest getting pylint/flake8 set up in VSCode so that it's easier to catch these sorts of issues

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+
+        #Finds tarball that matches GenerateFilename function output
+        os.chdir(f'{probe_dir}/staged/store')
+        for tarball in glob.glob("tz.*.test-extension__*"): 
```
This is a good catch that the glob may return multiple values. Since the function should NOT create multiple tarballs, you should add an assertion here that you only get one tarball back

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        file_obj= tarfile.open(tarball,"r")
+        namelist=file_obj.getnames()
+        for names in namelist:
+            return(names)
```
Same comment about a strange return in here

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        namelist=file_obj.getnames()
+        for names in namelist:
+            return(names)
+        file_obj.close()
```
I think you can avoid the `close()` here if you use a `with` wrapper

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        names.sort()
+        outfiles.sort()
+
+        self.assertListEqual(names, outfiles)
```
This is a good assertion! Another one would be to also check that the contents are expected

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,120 @@
+        # add content to the files
+        for fname in self.outfiles:
+            with open(os.path.join(self.outbox, fname), 'w') as f:
+                f.write('test content')
```
It would be good to make the contents specific per file (perhaps insert the filename at the front?)

brianhlin reviewed (changes requested):
A couple of improvements to be made but this is looking great!

jeff-takaki committed:
Comment Add Whitespace

Co-authored-by: Brian Lin <brianhlin@gmail.com>

jeff-takaki committed:
Add more assertions and move patches to setUp
Make corrections to tarball file extraction and tarball location check

jeff-takaki commented:
@brianhlin Made a few changes per your suggestions.

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,137 @@
+        self.assertTrue(os.path.exists(tarball_location),
+                        'Tarball not created in correct location')
+        self.assertEqual(tarball_count, 1,
+                         'Tarball created in directory != 1')
```
```suggestion
                         f'Expected 1 tarball, found {tarball_count}')
```

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,137 @@
+
+        # Finds exactly one tarball that matches GenerateFilename function output
+        tarball = glob.glob("tz.*.test-extension__*", root_dir=path_to_tarball)
+        tarball_location = os.path.join(f'{path_to_tarball}', tarball[0])
```
What happens when you find 0 tarballs?

brianhlin reviewed (approved):
Looks solid! A suggestion to improve the message upon one test failure and a slight logic issue. Pre-approving

jeff-takaki committed:
Fix test failure message for tarball count.

Co-authored-by: Brian Lin <brianhlin@gmail.com>

brianhlin commented on test/test_sandbox_mgmt.py:
```diff
@@ -0,0 +1,137 @@
+        # Finds tarball that matches GenerateFilename() output
+        tarball = glob.glob("tz.*.test-extension__*", root_dir=path_to_tarball)
+        # Where tarball exists
+        tarball_location = os.path.join(f'{path_to_tarball}', tarball[0])
```
I just noticed: this will also blow up if there are 0 tarballs

jeff-takaki committed:
Refactor tarball locator logic
get_tarball_function handles IndexError exception and fails test if no tarball is created.
test_tarball_creation and test_tarball_contents both make a call to this function

jeff-takaki commented:
@brianhlin Refactored!
